comm/src/main/scala/coop/rchain/comm/transport/StreamHandler.scala

```diff
@@ -51,30 +56,51 @@ object StreamHandler {
 
   private def collect(
       init: Streamed,
-      stream: Observable[Chunk]
-  ): EitherT[Task, Throwable, Streamed] =
-    EitherT(
-      (stream
-        .foldLeftL(init) {
-          case (stmd, Chunk(Chunk.Content.Header(ChunkHeader(sender, typeId, compressed, cl)))) =>
-            stmd.copy(
-              sender = sender.map(ProtocolHelper.toPeerNode(_)),
-              typeId = Some(typeId),
-              compressed = compressed,
-              contentLength = Some(cl)
-            )
-          case (stmd, Chunk(Chunk.Content.Data(ChunkData(newData)))) =>
-            stmd.fos.write(newData.toByteArray)
-            stmd.fos.flush()
-            stmd
-        })
-        .attempt
-    )
+      stream: Observable[Chunk],
+      circuitBreaker: CircuitBreaker
+  ): EitherT[Task, Throwable, Streamed] = {
+
+    def collectStream = stream.foldWhileLeftL(init) {
+      case (stmd, Chunk(Chunk.Content.Header(ChunkHeader(sender, typeId, compressed, cl)))) =>
+        Left(
+          stmd.copy(
+            sender = sender.map(ProtocolHelper.toPeerNode(_)),
+            typeId = Some(typeId),
+            compressed = compressed,
+            contentLength = Some(cl)
+          )
+        )
+      case (stmd, Chunk(Chunk.Content.Data(ChunkData(newData)))) =>
+        val array = newData.toByteArray
+        stmd.fos.write(array)
+        stmd.fos.flush()
+        val readSoFar = stmd.readSoFar + array.length
+        if (circuitBreaker(readSoFar))
+          Right(stmd.copy(circuitBroken = true))
+        else
+          Left(stmd.copy(readSoFar = readSoFar))
+    }
+
+    EitherT(collectStream.attempt.map {
+      case Right(stmd) if stmd.circuitBroken => Left(new RuntimeException("Circuit was broken"))
```

Comment:
Please remove the file when the circuit is broken. Also remove the file in case of `Left`.

Reply:
👍